packages/workbox-background-sync/lib/StorableRequest.mjs

```diff
@@ -0,0 +1,116 @@
+/*
+ Copyright 2017 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+const serializableProperties = [
+  'method',
+  'body',
+  'referrer',
+  'referrerPolicy',
+  'mode',
+  'credentials',
+  'cache',
+  'redirect',
+  'integrity',
+  'keepalive',
+  'signal',
+];
+
+
+/**
+ * A class to make it easier to serialize and de-serialize requests so they
+ * can be stored in IndexedDB.
+ */
+export default class StorableRequest {
+  /**
+   * Converts a Request object to a plain object that can be structured
+   * cloned or JSON-stringified.
+   *
+   * @param {Request} request
+   * @return {Promise<StorableRequest>}
+   */
+  static async fromRequest(request) {
+    const requestInit = {headers: {}};
+
+    // Set the body, if used.
+    if (request.bodyUsed) {
+      requestInit.body = await request.text();
+    }
+
+    // Convert the headers from an iterable to an object.
+    for (const [key, value] of request.headers.entries()) {
+      requestInit.headers[key] = value;
+    }
+
+    // Add all other serializable request properties
+    for (const prop of serializableProperties) {
+      if (request[prop] != null) {
+        requestInit[prop] = request[prop];
+      }
+    }
+
+    return new StorableRequest({url: request.url, requestInit});
+  }
+
+  /**
+   * Accepts a URL and RequestInit dictionary that can be used to create a
+   * new Request object. A timestamp is also set in the event that when this
+   * object was created is relevant after it's stored.
+   *
+   * @param {StorableRequest|Object} param1
```

Comment:
I don't follow this JSDoc—what's `{StorableRequest|Object}`?

Reply:
It means it could be of either type. In this case a `StorableRequest` object could be created from a plain object or "cloned" from another `StorableRequest` instance.

However, looking at the code, I think I only create this from plain objects, so I can remove.